### ISSUE TITLE
[ComfyUI v0.23.0] Sampling hangs at 0/N, GPU pinned 100%, with PromptRelay (LTX-2.3, Windows) — works fine on v0.22.0

### ISSUE THREAD

**Issue by Airsiren** [comfyanonymous/ComfyUI#14253]
### Custom Node Testing

- [ ] I have tried disabling custom nodes and the issue persists (see [how to disable custom nodes](https://docs.comfy.org/troubleshooting/custom-node-issues#step-1%3A-test-with-all-custom-nodes-disabled) if you need help)

## Summary
  A custom node (kijai/ComfyUI-PromptRelay) injects an additive attention bias into LTX-2.3 cross-attention using the public
  extension mechanism — `ModelPatcher.add_object_patch()` on each block's cross-attn `forward`, plus
  `transformer_options["optimized_attention_override"]` to route mask-bearing calls to
  `comfy.ldm.modules.attention.attention_pytorch`. This works on **ComfyUI v0.22.0** (~500s) but **hangs indefinitely on v0.23.0**:
  sampling sits at `0/N` with the GPU pinned at 100% and never advances. Reverting ComfyUI to v0.22.0 fixes it completely.

  Filed downstream at kijai/ComfyUI-PromptRelay#18; opening here in case it's a core-side regression in the object_patch /
  attention-override / masked-attention path.

## Environment
  - ComfyUI: **0.23.0** (deps: comfy-aimdo 0.4.7, comfy-kitchen 0.2.10, frontend 1.44.19)
  - OS: Windows 11
  - GPU: RTX 5090 32GB (Blackwell), driver 595.79
  - Python 3.11.8, PyTorch 2.10.0+cu128, CUDA 12.8
  - Attention: SageAttention 2.2.0.post4, launched with `--use-sage-attention --disable-dynamic-vram`
  - Model: LTX-2.3 (LTXAV) I2V, 1920x1088, 511 pixel frames (64 latent frames), 8 steps

 

## Symptom (log)
  Models load fine, then:
  [PromptRelay] Built penalty matrix (scaled): Lq=34170, Lk=1024, nonzero=13041243/34990080
  [PromptRelay] Built penalty matrix (scaled): Lq=426, Lk=1024, nonzero=162567/436224
  0%|          | 0/8 [00:00<?, ?it/s]
  …then it sits at `0/8` indefinitely with GPU at 100% (not a deadlock — it computes but never advances). Model loaded partially:
  `loaded partially; 23237 MB loaded, 799 MB offloaded, lowvram patches: 103`.

### Debug Logs

```powershell
## What I've ruled out
  - **Not pinned memory (#14250)**: models load completely (no `HostBuffer.read_file_slice` error); `--disable-pinned-memory` does
  **not** help.
  - **Not the new SDPA backend**: v0.23.0 added a Windows cuDNN-first SDPA priority in `comfy/ops.py`, and `attention_pytorch` now
  routes through `comfy.ops.scaled_dot_product_attention`. Monkey-patching `comfy.ops.scaled_dot_product_attention` back to plain
  `torch.nn.functional.scaled_dot_product_attention` (bypassing the cuDNN-first priority) — **still hangs**. So the cuDNN backend is
  not the cause.

  ## Possible cause (core side)
  v0.23.0 made large changes to `comfy/model_patcher.py` (`add_object_patch` / `get_model_object`), `comfy/ops.py`, and
  `comfy/model_management.py` (loader/offload). The custom node's pattern is:
  `add_object_patch("diffusion_model.transformer_blocks.{i}.attn2.forward", wrapper)`, where the wrapper builds an additive
  `[Lq≈34170, Lk≈1024]` mask, sets `optimized_attention_override`, and calls the underlying forward; the override routes mask-bearing
  calls to `attention_pytorch`. Worked on v0.22.0, hangs on v0.23.0. Did the semantics of `add_object_patch` / `get_model_object`,
  or the masked `attention_pytorch` path, change in v0.23.0?

  ## Caveat
  I have **not** isolated a minimal repro without the custom node (stock LTX-2.3), so I can't yet confirm whether this is a core
  regression or a custom-node compatibility issue. Happy to provide the full workflow JSON / startup log, or test any patch.
```

**Comment by Airsiren** [comment]
**Root cause found: the v0.23.0 async weight offloader, not a custom-node issue**

  Update after isolating this on a dedicated test box (RTX 4090 24GB, same software stack). The "hang" is actually **catastrophic
  slowdown**, not a deadlock.

  **Localization (py-spy):** time is spent in `comfy/ldm/lightricks/av_model.py` → `get_ada_values` / `expand_for_computation` (AdaLN
  `scale_shift_table.to(device)`), called many times per transformer block. This file is byte-identical between v0.22.0 and v0.23.0,
  so the model code is not the cause.

  **Decisive A/B (same workflow, 768×512, identical offload volume ~2.9GB):**

  | async offloader | step time |
  |---|---|
  | on (default) | **172 s/it** |
  | `--disable-async-offload` | **19.9 s/it** |

  → **8.6× slowdown from the async weight offloader alone**, at the same amount of offloaded weights. The LTX-2.3 AV model's
  high-frequency small-tensor `.to(device)` access pattern appears pathological for the async-stream offload path.

  **Workaround:** launch with `--disable-async-offload` (verified at reduced resolution; full-res on 24GB OOM-crashed, untested on
  ≥32GB).

  Happy to provide the full py-spy dumps / logs.

**Comment by Airsiren** [comment]
### Status update: still unfixed as of v0.24.1, requesting triage

  Re-checked on 2026-06-12. The async-offloader regression described above is **still present and unaddressed in released versions**:

  - **v0.24.0 / v0.24.1 do not fix it.** I went through the v0.24.0 PR list: the only offloading-related entry is #14218 ("old
  offloading mode"), which only touches the triposplat preview path. There is **no change to `comfy/model_management.py` async
  offload, `comfy/ops.py` cast/offload, or the LTX `av_model.py` AdaLN path** in either release.
  - Adjacent recent work (#14410 / CORE-288, and #14383) is in the same `ops`/loader area but targets the pinned-memory
  `HostBuffer.truncate` path, **not** this AdaLN high-frequency `.to(device)` async-stream slowdown.

  **Recap of the localized root cause** (unchanged):
  - py-spy puts the time in `comfy/ldm/lightricks/av_model.py` → `get_ada_values` / `expand_for_computation` (AdaLN
  `scale_shift_table.to(device)`), called many times per transformer block. This file is **byte-identical between v0.22.0 and
  v0.23.0**, so the model code is not the cause.
  - Same-offload-volume A/B (768×512, ~2.9 GB offloaded): async offloader **on = 172 s/it**, `--disable-async-offload` = **19.9
  s/it** → **8.6× slowdown from the offloader alone**.

  **Question for maintainers:** is the async weight offloader expected to handle a high-frequency small-tensor `.to(device)` access
  pattern like LTX-2.3's AdaLN, or is this a known pathological case? Is a fix planned for the async-stream path?

  **Caveat (unchanged) + offer:** my A/B workflow still includes the PromptRelay custom node; I have not yet built a stock-LTX-2.3
  minimal repro. If a clean stock repro (no custom nodes) is what's needed to triage this as a core regression, I can produce one —
  just confirm and I'll prioritize it.

  Environment: ComfyUI v0.23.0+, Windows 11, RTX 5090 32GB (also reproduced on 4090 24GB), Python 3.11.8, PyTorch 2.10.0+cu128,
  SageAttention 2.2.0.post4, `--use-sage-attention --disable-dynamic-vram`.

**Comment by Airsiren** [comment]
### Update: v0.25.0 makes this worse — the `--disable-async-offload` workaround no longer recovers speed

  Tested v0.25.0 (released 2026-06-16, commit 135abed8) on a dedicated RTX 4090 24GB box, same software stack, same workflow,
  768×512, ~2.9 GB offloaded — identical conditions to my earlier v0.23.0 A/B.

  | Config | v0.25.0 (today) | v0.23.0 (earlier) |
  |--------|-----------------|-------------------|
  | async offloader **on** (default) | **177 s/it** | 172 s/it |
  | `--disable-async-offload` | **~108 s/it** (steady over 3 steps) | **19.9 s/it** |
  | `--disable-async-offload --disable-pinned-memory` | **190 s/it** (worse) | — |

  **Key regression:** on v0.23.0, `--disable-async-offload` fully mitigated this (8.6× → ~20 s/it, usable). On v0.25.0 the same flag
  only gets ~108 s/it — still ~5.4× slower than the v0.23.0 workaround. So v0.25.0 hasn't just left the async-offload regression
  unfixed; **the synchronous offload path itself is now slow too**, and the workaround that used to rescue LTX-2.3 no longer does.

  - Disabling pinned memory makes it *worse* (190 s/it), so pinned memory is not the culprit — it was helping. The slowdown is in the
  weight-cast/offload path during the LTX AdaLN `.to(device)` access (py-spy localization unchanged from above).
  - v0.25.0's offload-area changes (#14480 comfy-aimdo 0.4.10 + `--reserve-vram`/`--vram-headroom`, #14437 `--high-ram`, #14390
  cudnn.benchmark) do not touch the `av_model.py` AdaLN async path, and appear to have slowed the non-async path as a side effect.

  **Net:** for LTX-2.3 on Windows, v0.25.0 is a regression vs v0.23.0 (no fix + degraded workaround). v0.22.0 remains the only
  version that runs this at full speed.

  Caveat (unchanged): my workflow still includes the PromptRelay custom node; I have not yet isolated a stock-LTX-2.3 minimal repro.
  Happy to produce one if that's what's needed to triage.

  Environment: ComfyUI v0.25.0, Windows 11, RTX 4090 24GB (production target is RTX 5090 32GB on v0.22.0), Python 3.11.8, PyTorch
  2.10.0+cu128, SageAttention 2.2.0.post4, comfy-aimdo 0.4.10.
